worker/src/main/java/gov/cms/ab2d/worker/processor/coverage/CoverageDriverImpl.java

```diff
@@ -0,0 +1,190 @@
+
+        for (Contract contract : attestedContracts) {
+            ZonedDateTime attestationTime = contract.getESTAttestationTime();
+
+            // Force first coverage period to be after
+            // January 1st 2020 which is the first moment we report data for
+            if (attestationTime.isBefore(epoch)) {
+                log.debug("contract attested before ab2d epoch setting to epoch");
+                attestationTime = getAB2DEpoch();
+            }
+
+            int coveragePeriodsForContracts = 0;
+            while (attestationTime.isBefore(now)) {
+                coverageService.getCreateIfAbsentCoveragePeriod(contract, attestationTime.getMonthValue(), attestationTime.getYear());
+                coveragePeriodsForContracts += 1;
+
+                attestationTime = attestationTime.plusMonths(1);
+            }
+
+            log.info("discovered {} coverage periods for contract {}", coveragePeriodsForContracts,
+                    contract.getContractName());
+        }
+
+        log.info("discovered all coverage periods now exiting");
+    }
+
+    private Set<CoveragePeriod> findAndCancelStuckCoverageJobs() {
+
+        log.info("attempting to find all stuck coverage searches and then cancel those stuck coverage searches");
+
+        Set<CoveragePeriod> stuckJobs = new LinkedHashSet<>(
+                coverageService.coveragePeriodStuckJobs(OffsetDateTime.now(ZoneOffset.UTC).minusHours(config.getStuckHours())));
+
+        for (CoveragePeriod period : stuckJobs) {
+            coverageService.failSearch(period.getId(), "coverage period current job has been stuck for at least "
+                    + config.getStuckHours() + " hours and is now considered failed.");
+        }
+
+        return stuckJobs;
+    }
+
+    private Set<CoveragePeriod> findStaleCoverageInformation() {
+
+        log.info("attempting to find all coverage information that is out of date and reduce down to coverage periods");
+
+        Set<CoveragePeriod> stalePeriods = new LinkedHashSet<>();
+        int monthsInPast = 0;
+        OffsetDateTime dateTime = OffsetDateTime.now(DateUtil.AB2D_ZONE);
+
+        do {
+            // Get past month and year
+            OffsetDateTime pastMonthTime = dateTime.minusMonths(monthsInPast);
+            int month = pastMonthTime.getMonthValue();
+            int year = pastMonthTime.getYear();
+
+            // Look for coverage periods that have not been updated
+            long daysInPast = config.getStaleDays() * (monthsInPast + 1);
+            OffsetDateTime lastUpdatedAfter = dateTime.minusDays(daysInPast);
+
+            stalePeriods.addAll(coverageService.coveragePeriodNotUpdatedSince(month, year, lastUpdatedAfter));
+            monthsInPast++;
+        } while (monthsInPast < config.getPastMonthsToUpdate());
+
+        return stalePeriods;
+    }
+
+    @Scheduled(fixedDelay = SIXTY_SECONDS, initialDelayString = "${coverage.update.initial.delay}")
```

Comment:
I love the Scheduled annotation but by itself, will run on every machine in the cluster.  That is why there are all those funky Quartz configuration files lying around the project.

Is it ok if this runs on mulitple nodes?  If yes, please add a comment.

If not, need to guarantee that.

Reply:
It is okay that it runs on every machine. This is how we balance the work between nodes.